File: pkg/EarthSciASTDiff.jl/src/cse.jl
# Hash-consed common-subexpression extraction over emitted band coefficients
# — the template half of the coefficient-size control roadmap item.
#
# Repeated coefficient subtrees (across bands, and repeated occurrences
# within one band) are hoisted into zero-parameter, match-less
# `expression_templates` entries carried INSIDE the `jacobians` block
# (esm-jacobian-spec.md §4); each occurrence is replaced by an
# `apply_expression_template` reference. This reuses the format's single
# structural-substitution mechanism (esm-spec §9.6) rather than inventing a
# new sharing construct: a zero-param template is a fixed named fragment,
# references may form a DAG (template body referencing an earlier template),
# and expansion is pure syntactic substitution — so
# [`expand_templates`](@ref) reproduces the original trees exactly.
#
# Free index names inside a hoisted fragment (band row indices, aggregate
# output indices) stay free in the template body and re-bind at the
# reference site after expansion — deliberate capture, since extraction is
# exact subtree factoring, not hygienic abstraction.

# ── hash-consing ─────────────────────────────────────────────────────────────
#
# Bottom-up interning: each structurally distinct subtree gets an integer id.
# A node's key is its serialized form with every child replaced by a
# placeholder carrying the child's id — O(node) per node instead of the
# quadratic full-subtree serialization.

struct _Intern
    ids::Dict{String,Int}    # shell key → id
    reps::Vector{ASTExpr}    # representative subtree per id
    nnodes::Vector{Int}      # full subtree node count per id
    counts::Vector{Int}      # total occurrences over everything interned
end
_Intern() = _Intern(Dict{String,Int}(), ASTExpr[], Int[], Int[])

_placeholder(id::Int) = VarExpr("__cse⟦$id⟧")

# (shell key, node count) of `e` given already-interned children.
function _shell_key(e::ASTExpr, childids::Vector{Int}, it::_Intern)
    if e isa OpExpr
        i = Ref(0)
        shell = map_children(_ -> (i[] += 1; _placeholder(childids[i[]])), e)
        nn = 1
        for cid in childids
            nn += it.nnodes[cid]
        end
        return stable_json(EarthSciAST.serialize_expression(shell)), nn
    end
    return stable_json(EarthSciAST.serialize_expression(e)), 1
end

function _intern!(it::_Intern, e::ASTExpr)::Int
    childids = Int[]
    e isa OpExpr && map_children(c -> (push!(childids, _intern!(it, c)); c), e)
    key, nn = _shell_key(e, childids, it)
    id = get!(it.ids, key) do
        push!(it.reps, e); push!(it.nnodes, nn); push!(it.counts, 0)
        length(it.reps)
    end
    it.counts[id] += 1
    return id
end

# Rewrite: replace every subtree whose id is `target` by `repl`, recomputing
# original ids bottom-up against the SAME (read-only) table. Returns
# (rewritten node, ORIGINAL id of `e`).
function _replace_id(e::ASTExpr, it::_Intern, target::Int, repl::ASTExpr)
    childids = Int[]; newchildren = ASTExpr[]
    if e isa OpExpr
        map_children(e) do c
            nc, cid = _replace_id(c, it, target, repl)
            push!(childids, cid); push!(newchildren, nc)
            c
        end
    end
    key, _ = _shell_key(e, childids, it)
    id = it.ids[key]
    id == target && return (repl, id)
    if e isa OpExpr && !isempty(newchildren)
        i = Ref(0)
        e = map_children(_ -> (i[] += 1; newchildren[i[]]), e)
    end
    return (e, id)
end

# ── extraction ───────────────────────────────────────────────────────────────

"""
    cse_templates(exprs; min_nodes = 12, prefix = "jt")
        -> (templates::OrderedDict{String,ASTExpr}, rewritten::Vector{ASTExpr})

Greedy largest-first extraction of repeated subtrees over `exprs`. Each
round interns the whole working set (input expressions plus the bodies of
templates already extracted), picks the largest subtree that occurs at least
twice and spans at least `min_nodes` AST nodes, names it `"\$(prefix)\$k"`,
and replaces every occurrence — in the inputs AND in earlier template
bodies — with an `apply_expression_template` reference. Terminates when no
candidate remains; the result is a reference DAG (acyclic by strict
containment). Deterministic: ties between equally-sized candidates break by
interning (first-encounter) order over the given expression order.
"""
function cse_templates(exprs::Vector{ASTExpr}; min_nodes::Int = 12,
                       prefix::String = "jt")
    work = ASTExpr[exprs...]
    nin = length(work)
    names = String[]
    while true
        it = _Intern()
        for e in work
            _intern!(it, e)
        end
        best = 0
        for id in 1:length(it.reps)
            it.counts[id] >= 2 || continue
            it.nnodes[id] >= min_nodes || continue
            r = it.reps[id]
            r isa OpExpr || continue
            r.op == "apply_expression_template" && continue
            (best == 0 || it.nnodes[id] > it.nnodes[best]) && (best = id)
        end
        best == 0 && break
        name = "$(prefix)$(length(names) + 1)"
        push!(names, name)
        repl = OpExpr("apply_expression_template", ASTExpr[]; name = name)
        body = it.reps[best]
        work = ASTExpr[_replace_id(e, it, best, repl)[1] for e in work]
        push!(work, body)   # the body joins the pool: later rounds may factor it
    end
    templates = OrderedDict{String,ASTExpr}(
        n => work[nin + k] for (k, n) in enumerate(names))
    return templates, work[1:nin]
end

"""
    expand_templates(e::ASTExpr, registry) -> ASTExpr

Inverse of [`cse_templates`](@ref): substitute every zero-parameter
`apply_expression_template` reference by its body from `registry`
(name → body `ASTExpr`), recursively (the registry is a DAG). Throws
`ArgumentError` on a reference to a name the registry does not contain.
"""
function expand_templates(e::ASTExpr, registry::AbstractDict{String,ASTExpr})::ASTExpr
    e isa OpExpr || return e
    if e.op == "apply_expression_template"
        e.name !== nothing && haskey(registry, e.name) || throw(ArgumentError(
            "apply_expression_template reference `$(e.name)` not in the block's " *
            "expression_templates"))
        return expand_templates(registry[e.name], registry)
    end
    return map_children(x -> expand_templates(x, registry), e)
end

File: pkg/earthsci-astdiff-py/src/earthsci_astdiff/cse.py
"""Hash-consed template CSE over emitted band coefficients (cse.jl).

Deterministic and byte-compatible with the Julia reference: same interning
(first-encounter) order, same largest-first tie-break, same `jt` naming —
the cross-runtime block goldens pin it.
"""
from __future__ import annotations

from earthsci_ast.esm_types import Expr, ExprNode
from earthsci_ast.expr_walk import map_children

from .expr_helpers import ser_expr, stable_json


class _Intern:
    def __init__(self):
        self.ids: dict[str, int] = {}
        self.reps: list[Expr] = []
        self.nnodes: list[int] = []
        self.counts: list[int] = []

    def intern(self, e: Expr) -> int:
        childids: list[int] = []
        if isinstance(e, ExprNode):
            map_children(e, lambda c: (childids.append(self.intern(c)), c)[1])
        key, nn = self._shell_key(e, childids)
        got = self.ids.get(key)
        if got is None:
            self.reps.append(e)
            self.nnodes.append(nn)
            self.counts.append(0)
            got = self.ids[key] = len(self.reps)
        self.counts[got - 1] += 1
        return got

    def _shell_key(self, e: Expr, childids: list[int]) -> tuple[str, int]:
        if isinstance(e, ExprNode):
            i = [0]

            def repl(_c: Expr) -> Expr:
                i[0] += 1
                return f"__cse⟦{childids[i[0] - 1]}⟧"

            shell = map_children(e, repl)
            nn = 1 + sum(self.nnodes[cid - 1] for cid in childids)
            return stable_json(ser_expr(shell)), nn
        return stable_json(ser_expr(e)), 1


def _replace_id(e: Expr, it: _Intern, target: int, repl: Expr):
    """Replace every subtree whose id is ``target``; ids recomputed bottom-up
    against the read-only table. Returns (rewritten, original id of e)."""
    childids: list[int] = []
    newchildren: list[Expr] = []
    if isinstance(e, ExprNode):
        def visit(c: Expr) -> Expr:
            nc, cid = _replace_id(c, it, target, repl)
            childids.append(cid)
            newchildren.append(nc)
            return c

        map_children(e, visit)
    key, _ = it._shell_key(e, childids)
    eid = it.ids[key]
    if eid == target:
        return repl, eid
    if isinstance(e, ExprNode) and newchildren:
        i = [0]

        def take(_c: Expr) -> Expr:
            i[0] += 1
            return newchildren[i[0] - 1]

        e = map_children(e, take)
    return e, eid


def cse_templates(exprs: list[Expr], min_nodes: int = 12,
                  prefix: str = "jt"):
    """Greedy largest-first extraction of repeated subtrees. Returns
    (templates: {name: body}, rewritten exprs)."""
    work = list(exprs)
    nin = len(work)
    names: list[str] = []
    while True:
        it = _Intern()
        for e in work:
            it.intern(e)
        best = 0
        for i in range(1, len(it.reps) + 1):
            if it.counts[i - 1] < 2 or it.nnodes[i - 1] < min_nodes:
                continue
            r = it.reps[i - 1]
            if not isinstance(r, ExprNode):
                continue
            if r.op == "apply_expression_template":
                continue
            if best == 0 or it.nnodes[i - 1] > it.nnodes[best - 1]:
                best = i
        if best == 0:
            break
        name = f"{prefix}{len(names) + 1}"
        names.append(name)
        repl = ExprNode(op="apply_expression_template", args=[], name=name)
        body = it.reps[best - 1]
        work = [_replace_id(e, it, best, repl)[0] for e in work]
        work.append(body)          # later rounds may factor the body itself
    templates = {n: work[nin + k] for k, n in enumerate(names)}
    return templates, work[:nin]


def expand_templates(e: Expr, registry: dict[str, Expr]) -> Expr:
    if not isinstance(e, ExprNode):
        return e
    if e.op == "apply_expression_template":
        if e.name is None or e.name not in registry:
            raise ValueError(
                f"apply_expression_template reference `{e.name}` not in the "
                "block's expression_templates")
        return expand_templates(registry[e.name], registry)
    return map_children(e, lambda x: expand_templates(x, registry))
